trade-finance-manager-api/src/v1/controllers/auth/auth-provider.js

```diff
@@ -0,0 +1,183 @@
+       */
+      scopes: [],
+    };
+
+    const authCodeRequestParams = {
+      state,
+
+      /**
+       * By default, MSAL Node will add OIDC scopes to the auth code request. For more information, visit:
+       * https://docs.microsoft.com/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes
+       */
+      scopes: [],
+    };
+
+    /**
+     * If the current msal configuration does not have cloudDiscoveryMetadata or authorityMetadata, we will
+     * make a request to the relevant endpoints to retrieve the metadata. This allows MSAL to avoid making
+     * metadata discovery calls, thereby improving performance of token acquisition process.
+     */
+
+    if (!this.config.msalConfig.auth.authorityMetadata) {
+      const authorityMetadata = await this.getAuthorityMetadata();
+      this.config.msalConfig.auth.authorityMetadata = JSON.stringify(authorityMetadata);
+    }
+    const msalInstance = this.getMsalInstance(this.config.msalConfig);
+
+    // trigger the first leg of auth code flow
+    return this.getAuthCodeUrl(
+      csrfToken,
+      authCodeUrlRequestParams,
+      authCodeRequestParams,
+      msalInstance,
+    );
+  }
+
+  /**
+   * Prepares the auth code request parameters and initiates the first leg of auth code flow
+   * @param req: Express request object
+   * @param res: Express response object
+   * @param next: Express next function
+   * @param authCodeUrlRequestParams: parameters for requesting an auth code url
+   * @param authCodeRequestParams: parameters for requesting tokens using auth code
+   */
+  async getAuthCodeUrl(csrfToken, authCodeUrlRequestParams, authCodeRequestParams, msalInstance) {
+    // Generate Proof Key of Code Exchange (PKCE) Codes before starting the authorization flow
+    const { verifier, challenge } = await this.cryptoProvider.generatePkceCodes();
+    // Set generated PKCE codes and method as session vars
+    const response = { csrfToken };
+    response.pkceCodes = {
+      challengeMethod: 'S256',
+      verifier,
+      challenge,
+    };
+    /**
+     * By manipulating the request objects below before each request, we can obtain
+     * auth artifacts with desired claims. For more information, visit:
+     * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationurlrequest
+     * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationcoderequest
+     * */
+    response.authCodeUrlRequest = {
+      ...authCodeUrlRequestParams,
+      redirectUri: this.config.redirectUri,
+      responseMode: 'form_post', // recommended for confidential clients
+      codeChallenge: response.pkceCodes.challenge,
+      codeChallengeMethod: response.pkceCodes.challengeMethod,
+    };
+
+    response.authCodeRequest = {
+      ...authCodeRequestParams,
+      redirectUri: this.config.redirectUri,
+      code: '',
+    };
```

Comment:
Could simplify some of this with destructuring and maybe less `reponse.property = { ...` 🤔